### DISCUSSION TITLE
How does pipux differ from 'uv tool'?

### DISCUSSION THREAD

**Discussion by kwaegel** [bulletmark/pipxu#11] [General]
What is the difference between `pipux` and `uv tool` right now?

It looks like UV has support for installing tools now via `uv tool install ...`, and it would be helpful for people coming from `pipx` (like myself) to understand why I would pick one over the other.

From cursory inspection, it looks like they're both trying to do the same thing, but `uv tool` is less full-featured at the moment, and isn't trying for `pipx` compatibility.

**Comment by bulletmark** [comment]
I have somewhat been anticipating this question (BTW, note this tool is called `pipxu`, not `pipux`).

I created `pipxu` not long after `uv` was first introduced. At the time `uv` only provided `venv` and `pip` functionality and I had no idea at that time that the `uv` devs planned to add `pipx` functionality. I guess it could be argued that `uv tool` now supercedes `pipxu`.

BTW, `pipxu` is not really `pipx` compatible. It uses the same verbs for the main commands (i.e. `install`, `uninstall`, `upgrade`, `list`) but installs the app venvs in it's own directory. That last sentence is exactly true for `uv tool` as well. So `pipx`, `pipxu`, and `uv tool` all basically provide the same essential functionality, and with very similar commands.

I personally still use `pipxu` over `uv tool`. Speed wise, `pipxu` is written in Python but since it uses `uv` under the covers it is pretty much as fast as `uv tool` is. `pipxu` provides the `inject` and `uninject` commands like `pipx` has, but `uv tool` does not have.  As a developer I also use the ".' shortcut frequently which only `pipxu` has universally, and I also use the `pipxu` unique `debug` command frequently. 

As somewhat of an aside comment, I never implemented the `run` command in `pipxu` because I thought that command (and concept) was largely useless in `pipx` given how slow it was. However, `uv tool run` (i.e. `uvx`) installs and runs apps extremely quickly so I can really see the value in it mainly because naive users can simply follow an instruction, e.g. `uvx ruff check .` and not have to worry about installation or PATH issues etc. It is a universally easy way to install and run Python based apps on Linux, Mac, Windows etc.
